Feature/awf/angular cli

CloudNiner opened the pull request:
## Overview

Updates the projects Angular dependencies, to 4.0+, and replaces the custom webpack build with the Angular CLI, now that it is out of beta.

### Demo

Not really, app looks the same.


### Notes

The Angular CLI upgrade was relatively straightforward, but there was a lot of stuff to fix. I generated a new project outside the repo via `ng new` then copied the relevant CLI files into the project and deleted the old ones. Then cleaned up a bunch of changed file paths and broken stuff from the change.

## Testing Instructions

- `rm -rf node_modules`
- `npm install`
- `npm start`

The dev server should start and reload on file change. 

Test prod build:
- `npm run build:prod`
- `http-server -p 7777 dist/`

Load http://localhost:7777 should see the app working, with a much smaller total asset download (~1.2mb)

`npm run test` should complete successfully.

`npm run e2e` will fail because no e2e tests are defined.

`npm run lint` currently fails. ~~Going to spend some time attempting to clean this up as a separate commit on this PR later.~~ <-- EDIT: ~~The changes are massive and would detract from the changes made to update to Angular CLI. Will open as a separate PR.~~ See #143 


Closes #139, Closes #140

CloudNiner commented:
@fungjj92 I opened #143 to address the linting errors separately, should be able to look at these in either order, I'll just wait to merge both until they're approved.

fungjj92 commented:
Haven't played with the lab yet, but vagrant dies on provision

TASK [climate-change-lab.app : Production build] *******************************
fatal: [app]: FAILED! => {"changed": true, "cmd": ["npm", "run", "build:prod"], "delta": "0:00:50.802450", "end": "2017-06-09 14:13:32.736649", "failed": true, "rc": 1, "start": "2017-06-09 14:12:41.934199", "stderr": "You are running version 6.5.0 of Node, which will not be supported in future\nversions of the CLI. The official Node version that will be supported is 6.9 and greater.\n\nTo disable this warning use \"ng set --global warnings.nodeDeprecation=false\".

I bumped node to the LTS in group_vars/all, nodejs_version: 6.11.0

and that error went away, but more gunk, same step

CloudNiner commented:
Ok, I pushed a commit that fully revamps the Vagrant installation. You'll want to `vagrant destroy && vagrant up` for a full reset. Once it provisions, test:

- `npm start`
- `npm run build:prod`

Both should complete successfully.

fungjj92 commented:
Nice. With those changes, vagrant stuff runs cleanly.
 👍  prod build, and serving out /dist works via http-server

Outstanding process issues:
- `npm test` fails on stale reference, to angular2-ui-switch in `login.component.spec.ts` fixed with `import { UiSwitchModule } from 'ngx-ui-switch/src';`
- I'm still getting no response at localhost:4200, no errors. I'm guessing you don't run into this?

fungjj92 commented on README.md:
```diff
@@ -27,26 +27,27 @@ cd /opt/climate-change-lab
-| `npm run server:dev` | Serve project normally, without hot module replacement | If you don't want hot module replacement |
-| `npm run lint --silent` | Run typescript linter | To clean up your angular and .ts files |
+| `npm start` | Serve project in dev mode using Angular CLI | For quick browser updating with refresh |
+| `npm run lint` | Run typescript linter | To clean up your angular and .ts files |
```
This list would benefit from `npm test`, though prob not `npm run e2e` until/unless we have appropriate files

CloudNiner commented:
You're running `npm start` within Vagrant and don't see the lab when you hit localhost:4200 from the browser on your host machine? If so I know the fix for that, will update.

I'll add npm test and fixup the outstanding error.

CloudNiner commented:
`npm start` should now work as expected, and fixed tests. 

Note that tests will not actually run inside the VM, the test runner uses Chrome by default. 

I attempted to install phantomjs which failed to run tests as well, documented this separately on a new issue #145

fungjj92 commented:
While we're at removing stale files, ./src/assets/indicators.json and ./src/assets/mockdata/ and ./config/

After that, looking and working well for me

fungjj92 commented on tslint.json:
```diff
@@ -123,6 +107,29 @@
+    "directive-class-suffix": true,
+    "no-access-missing-member": true,
+    "templates-use-public": true,
+    "invoke-injectable": true
```
in all honesty i dont know what all these settings are but if they're defaults from ng-cli, that seems good to me since that's the standard convention

CloudNiner replied:
Yeah, its initialized with the default Angular Style Guide conventions when generated via Angular CLI. This tslint.json file is a combination of both [tslint](https://github.com/palantir/tslint) and the [Angular codelyzer](https://github.com/mgechev/codelyzer).

CloudNiner commented:
Good point about the stale files, I'll remove those before merging. 

Also waiting to merge until #143 is reviewed, which probably only needs a quick pass through and verification that `npm run lint` works as expected.

fungjj92 commented:
Spoke too soon, actually I am not seeing change detection when editing typescript or static files inside the VM.

CloudNiner commented:
Updated again. Angular CLI added a `--poll` option to `ng serve` that uses webpack polling to detect file changes rather than relying on pushes from the filesystem (the default, and faster). Adding this change saw the dev server consistently picking up file changes in `.ts` and `.html` files within 2-4s. Also:
- Added a bunch of clarifying language on when to use the VM
- Cleaned up the Vagrantfile definition, see the "modernize project Vagrantfile" commit.

fungjj92 commented:
Nice, +1. Thanks for bearing with multi rounds of comments!
